packages/block-editor/src/components/inserter/template-parts.js

```diff
@@ -0,0 +1,197 @@
+				{ templatePart.slug }
+			</div>
+		</div>
+	);
+}
+
+function TemplatePartsByTheme( { templateParts, onInsert } ) {
+	const templatePartsByTheme = useMemo( () => {
+		return Object.values( groupBy( templateParts, 'meta.theme' ) );
+	}, [ templateParts ] );
+	const currentShownTPs = useAsyncList( templateParts );
+
+	return (
+		<>
+			{ templatePartsByTheme.length &&
+				templatePartsByTheme.map( ( templatePartList ) => (
+					<InserterPanel
+						key={ templatePartList[ 0 ].meta.theme }
+						title={ templatePartList[ 0 ].meta.theme }
+					>
+						{ templatePartList.map( ( templatePart ) => {
+							return currentShownTPs.includes( templatePart ) ? (
+								<TemplatePartItem
+									key={ templatePart.id }
+									templatePart={ templatePart }
+									onInsert={ onInsert }
+								/>
+							) : (
+								<TemplatePartPlaceholder
+									key={ templatePart.id }
+								/>
+							);
+						} ) }
+					</InserterPanel>
+				) ) }
+		</>
+	);
+}
+
+function TemplatePartSearchResults( { templateParts, onInsert, filterValue } ) {
+	const filteredTPs = useMemo( () => {
+		// Filter based on value.
+		const lowerFilterValue = filterValue.toLowerCase();
+		const searchResults = templateParts.filter(
+			( { slug, meta: { theme } } ) =>
+				slug.toLowerCase().includes( lowerFilterValue ) ||
+				theme.toLowerCase().includes( lowerFilterValue )
+		);
+		// Order based on value location.
+		searchResults.sort( ( a, b ) => {
+			// First prioritize index found in slug.
+			const indexInSlugA = a.slug
+				.toLowerCase()
+				.indexOf( lowerFilterValue );
+			const indexInSlugB = b.slug
+				.toLowerCase()
+				.indexOf( lowerFilterValue );
+			if ( indexInSlugA !== -1 && indexInSlugB !== -1 ) {
+				return indexInSlugA - indexInSlugB;
+			} else if ( indexInSlugA !== -1 ) {
+				return -1;
+			} else if ( indexInSlugB !== -1 ) {
+				return 1;
+			}
+			// Second prioritize index found in theme.
+			return (
+				a.meta.theme.toLowerCase().indexOf( lowerFilterValue ) -
+				b.meta.theme.toLowerCase().indexOf( lowerFilterValue )
+			);
+		} );
+		return searchResults;
+	}, [ filterValue, templateParts ] );
+
+	const currentShownTPs = useAsyncList( filteredTPs );
+
+	return (
+		<>
+			{ filteredTPs.map( ( templatePart ) => (
+				<InserterPanel
+					key={ templatePart.id }
+					title={ templatePart.meta.theme }
+				>
+					{ currentShownTPs.includes( templatePart ) ? (
+						<TemplatePartItem
+							key={ templatePart.id }
+							templatePart={ templatePart }
+							onInsert={ onInsert }
+						/>
+					) : (
+						<TemplatePartPlaceholder key={ templatePart.id } />
+					) }
+				</InserterPanel>
+			) ) }
+		</>
+	);
+}
+
+export default function TemplateParts( { onInsert, filterValue } ) {
+	const templateParts = useSelect( ( select ) => {
+		return select( 'core' ).getEntityRecords(
+			'postType',
+			'wp_template_part',
+			{
+				status: [ 'publish', 'auto-draft' ],
```

Comment:
Should we add

```suggestion
				resolved: true,
				status: [ 'publish', 'auto-draft' ],
				theme,
```

to only show template parts for the current theme? (This would mean we'd need to get `theme` via the `getCurrentThemeSelector`.)

Reply:
I don't think so.  The theme is arbitrary, a user created template parts will be saved under a theme tag that does not match the current theme.  Similarly, if we have the template part saved, the theme it originated from shouldn't be a restriction on if we can use it.

Reply:
Makes sense. Just to make sure though, could that be a problem when [exporting a theme](https://github.com/WordPress/gutenberg/pull/22922)?

Reply:
To put on my devil's advocate's hat, I'm not even 100% convinced that we should allow inserting template parts from a different theme TBH :thinking: E.g. I'd expect a theme's header template part to kind of reflect its overall look and feel, so using a different theme's template part might give a rather inconsistent look and feel.

Furthermore, if we make available the template parts from _all_ installed themes, it can be confusing to find the current theme's ones (e.g. to pick the right header). (Clearly, the latter is mitigated by grouping template parts by theme, as this PR does.)

Reply:
Template parts from non-active themes will only show if you have customized them.

Reply:
>  I'd expect a theme's header template part to kind of reflect its overall look and feel, so using a different theme's template part might give a rather inconsistent look and feel.

It definitely could, but it allows users to choose from other template parts they have created or customized.  If they want to use them over the theme's supplied header template part, they can choose to do so.  If the user still feels it doesn't fully match the style, they can be further customized.  

Maybe in the future we may need to take steps to make the template part supplied by the current theme to be more prominent (default to the top of the list or something). But since "Template parts from non-active themes will only show if you have customized them.", this may not be necessary.

Reply:
> Template parts from non-active themes will only show if you have customized them.

That could be confusing enough tho, no?

Reply:
> That could be confusing enough tho, no?

Good point, it could be confusing between base template parts supplied by the theme vs. ones you have customized from that theme?  In which case maybe we could end up presenting it as 'Customized from 'theme-name''. as opposed to just 'theme-name'

Reply:
I don't think we should show the non-customized ones if they have a customized counterpart.